### DISCUSSION TITLE
Implementing Secure Scuttlebutt (SSB) Protocols

### DISCUSSION THREAD

**Discussion by mycognosist** [libp2p/rust-libp2p#3456] [Q&A] [answered]
I'm playing around with libp2p in the context of a Scuttlebutt node. I was initially hoping to start a swarm with basic TCP connections to test connectivity, and then later tackle libp2p protocol implementations for [Secret Handshake](https://ssbc.github.io/scuttlebutt-protocol-guide/#peer-connections) and [MUXRPC](https://ssbc.github.io/scuttlebutt-protocol-guide/#rpc-protocol), but I don't think this is currently possible (based on https://github.com/libp2p/rust-libp2p/discussions/2396).

Reading further about some of the work that may be require to implement the SSB protocols in libp2p-land, I now wonder if libp2p would be viable. As I understand it, libp2p peers use [multistream-select](https://github.com/multiformats/multistream-select) for protocol negotiation. What if a node running libp2p wants to connect and communicate to a peer without libp2p (therefore no multistream-select)? Is that a non-starter? I'm thinking about interoperability with all the Scuttlebutt nodes which do not use libp2p.

Thanks for your patience with my clumsy question.

**Comment by thomaseizinger** [comment, marked as answer]
I am not too familiar with SSB but by the looks of it, it allows for an encrypted transport.

We do something similar with QUIC. There, the multiaddr (`/quic`) suggests that we set up a QUIC connection on the specified port without `multistream-select`.

I think SSB could be used to implement a `Transport` that offers authentication but no multiplexing, at least not in the same way libp2p applications expect it. Those want reliable, ordered streams on top of a transport.

You'd have to implement those on top of MUXRPC which seems a bit silly.

Do SSB nodes allow for something else other than MUXRPC to be used on top of the established connection?

**Reply by thomaseizinger** [reply]
Reading more into the docs, the use of MUXRPC seems to be implicit and therefore, there isn't really a way of running a different protocol.

I am afraid that makes two approaches incompatible on a fundamental level.

**Reply by mycognosist** [reply]
@thomaseizinger 

Thanks very much for taking the time to answer.
